The CTC decoder timesteps now corresponds to the timesteps of the most probable CTC path, instead of the earliest timesteps of all possible paths.

godefv opened the pull request:
This follows the issue #3180 .

I suggest a new way of handling timesteps produced by the CTC decoder. There is no strange heuristic, and I think the logic is clear : when fusing two different paths leading to the same prefix, not only fuse the probabilities (the probabilities are added), but also fuse the timestep sequences (for the last letter in the sequence, choose the timestep from the most probable path).

The place where two different paths leading to the same prefix are fused are the places where `log_sum_exp` is called, because this function fuses the probabilities. So, timesteps would now be fused at the same places.

The other change is that each `PathTrie` node would now store the full sequence of timesteps. This is because one prefix can be an ancestor of another and their timesteps on a given node can differ. Having the full sequence of timesteps in each node, we have no need to duplicate a node with different timesteps, and it is much simpler like that. 
Moreover, it makes sense to store the *full* sequence of timesteps, because the *combined* probabilities are also stored there. The total probability is not the sum of the probability of each output token, and, in the same way, the correct sequence of timesteps is not the concatenation of the timestep of each output token.

Since I need to compare the probability of different paths (to keep the timesteps of the most probable one), it is important to compare paths of the same length (eg. paths from the beginning up to the current time). So, exactly the same way as it is done for the probabilities, I need to know the timesteps of the previous time, and store the timesteps of the current time separately.

In the end, timesteps are handled in a way very similar to the way probabilities are handled.

### Results on an example
To evaluate the resulting timesteps, I first take the argmax of my logits. In my example, it gives :
```
tou_________________________________________________ss  les  _a__mouurreuux  de  se_p_ort__ diiivverrr_  ss'enn__ _r_é___j_uuii__rr__on_t____    aa_v_eecc_   l''aap__p_rroo___chhee_____    de_  ll'hhi__v_e_rr____  et   la   rre__p_rri_ssee  dee  la  c_ouppee ddu  mon_deee      ss__kk_i___      less    ii_mm_a__ggees_   de   _g_ll_i_ss__ssee____________________             ree__t_rrou_vveennt    uunee    __pllaa___cee____      de    _cchhooixx__  d_ans_  lles  _pp_a____ggeess        ssspp_o_r_t_ii_vees_  de  v_o_s_ _jourrnnaauxx  ttéé_l_é___v_ii___ss_é__s__      ddeeu_x_   __é___pprreeuu_vvees___________________          _auu_jjoouurrdd''hhuuii___       _o____nno___rr_o_____d__a___mm________ ___s___a_n______t__a______  _q__a___tt__e___rr_i____n__a_____     __p_r____mmie_r__  s__a___l__o___m___    _g_é____ant__  de   lla  _c_ou_ppee  ddu   m_on___deee____________________________________________________
```
As the logits are the only input of the decoder, I base my evaluation on them instead of comparing with the audio file directly. It is known that the CTC loss does not guarantee alignment between the audio file and the logits, so the best thing the decoder can do is to fit the logits as best as it can. This is reasonable because, in practice, the logits are aligned quite well with the audio file.

Then, for each word, I take the part of the logits corresponding to the output timesteps, take the argmax (as said above), and print the corresponding decoded text. 

Finally, I assume that good timesteps should lead to a good match between the word and its corresponding text decoded from the argmax of the logits.

Before this PR, the result in my example is (text between slashes is from the logits argmax, spaces are trimmed) : 
```
[WordScoreRange(word=tous /tou/, score=None, ranges=((0, 4),)),   
 WordScoreRange(word=les /les/, score=None, ranges=((55, 59),)),       
 WordScoreRange(word=amoureux /amoureu/, score=None, ranges=((60, 74),)),
 WordScoreRange(word=de /d/, score=None, ranges=((75, 78),)),           
 WordScoreRange(word=sport /seport/, score=None, ranges=((80, 90),)),           
 WordScoreRange(word=divers /diver/, score=None, ranges=((91, 102),)),
 WordScoreRange(word=s'en /s'en/, score=None, ranges=((103, 111),)),  
 WordScoreRange(word=réjouiront /réjuiron/, score=None, ranges=((112, 136),)),
 WordScoreRange(word=avec /avec/, score=None, ranges=((141, 153),)),  
 WordScoreRange(word=l'approche /l'approche/, score=None, ranges=((155, 179),)),
 WordScoreRange(word=de /de/, score=None, ranges=((185, 191),)),  
 WordScoreRange(word=l'hiver /l'hive/, score=None, ranges=((192, 206),)),
 WordScoreRange(word=et /e/, score=None, ranges=((208, 215),)),
 WordScoreRange(word=la /la/, score=None, ranges=((217, 221),)),
 WordScoreRange(word=reprise /reprise/, score=None, ranges=((222, 238),)),
 WordScoreRange(word=de /de/, score=None, ranges=((240, 243),)),
 WordScoreRange(word=la /la/, score=None, ranges=((244, 248),)),
 WordScoreRange(word=coupe /coupe/, score=None, ranges=((249, 257),)),
 WordScoreRange(word=du /d/, score=None, ranges=((258, 261),)),
 WordScoreRange(word=monde /monde/, score=None, ranges=((263, 270),)),
 WordScoreRange(word=de /e/, score=None, ranges=((271, 275),)),
 WordScoreRange(word=ski /ski/, score=None, ranges=((276, 286),)),
 WordScoreRange(word=les /les/, score=None, ranges=((290, 298),)),
 WordScoreRange(word=images /image/, score=None, ranges=((300, 316),)),
 WordScoreRange(word=de /d/, score=None, ranges=((318, 322),)),
 WordScoreRange(word=glisse /glisse/, score=None, ranges=((324, 341),)),
 WordScoreRange(word=retrouvent /retrouvent/, score=None, ranges=((363, 394),)),
 WordScoreRange(word=une /une/, score=None, ranges=((395, 402),)),
 WordScoreRange(word=place /place/, score=None, ranges=((404, 419),)),
 WordScoreRange(word=de /de/, score=None, ranges=((425, 432),)),
 WordScoreRange(word=choix /choix/, score=None, ranges=((433, 445),)),
 WordScoreRange(word=dans /dans/, score=None, ranges=((448, 455),)),
 WordScoreRange(word=les /les/, score=None, ranges=((457, 462),)),
 WordScoreRange(word=pages /pages/, score=None, ranges=((463, 478),)),
 WordScoreRange(word=sportives /sportives/, score=None, ranges=((479, 506),)),
 WordScoreRange(word=de /de/, score=None, ranges=((508, 511),)),
 WordScoreRange(word=vos /vos/, score=None, ranges=((512, 518),)),
 WordScoreRange(word=journaux /journaux/, score=None, ranges=((520, 532),)),
 WordScoreRange(word=télévisés /télévisé/, score=None, ranges=((533, 559),)),
 WordScoreRange(word=deux /deux/, score=None, ranges=((563, 575),)),
 WordScoreRange(word=épreuves /épreuve/, score=None, ranges=((577, 598),)),
 WordScoreRange(word=aujourd'hui /aujourd'hu/, score=None, ranges=((618, 649),)),
 WordScoreRange(word=on /o/, score=None, ranges=((654, 665),)),
 WordScoreRange(word=a /am/, score=None, ranges=((685, 699),)),
 WordScoreRange(word=santa /santa/, score=None, ranges=((703, 726),)),
 WordScoreRange(word=caterina /qaterina/, score=None, ranges=((729, 756),)),
 WordScoreRange(word=premier /prmier/, score=None, ranges=((762, 783),)),
 WordScoreRange(word=salon /salom/, score=None, ranges=((785, 803),)),
 WordScoreRange(word=géant /géant/, score=None, ranges=((805, 818),)),
 WordScoreRange(word=de /d/, score=None, ranges=((819, 823),)),
 WordScoreRange(word=la /l/, score=None, ranges=((825, 829),)),
 WordScoreRange(word=coupe /coupe/, score=None, ranges=((831, 841),)),
 WordScoreRange(word=du /d/, score=None, ranges=((842, 846),)),
 WordScoreRange(word=monde /mon/, score=None, ranges=((848, 857),))]
```

After this PR, the result in my example is : 
```
[WordScoreRange(word=tous /tous/, score=None, ranges=((0, 54),)), 
 WordScoreRange(word=les /les/, score=None, ranges=((56, 59),)),        
 WordScoreRange(word=amoureux /amoureux/, score=None, ranges=((62, 75),)),
 WordScoreRange(word=de /de/, score=None, ranges=((77, 79),)),          
 WordScoreRange(word=sport /seport/, score=None, ranges=((81, 91),)),           
 WordScoreRange(word=divers /diver/, score=None, ranges=((92, 103),)),
 WordScoreRange(word=s'en /s'en/, score=None, ranges=((105, 113),)),  
 WordScoreRange(word=réjouiront /réjuiront/, score=None, ranges=((114, 140),)),
 WordScoreRange(word=avec /avec/, score=None, ranges=((145, 155),)),  
 WordScoreRange(word=l'approche /l'approche/, score=None, ranges=((158, 185),)),
 WordScoreRange(word=de /de/, score=None, ranges=((189, 192),)),  
 WordScoreRange(word=l'hiver /l'hiver/, score=None, ranges=((194, 212),)),
 WordScoreRange(word=et /et/, score=None, ranges=((214, 216),)),                                                               
 WordScoreRange(word=la /la/, score=None, ranges=((219, 221),)),
 WordScoreRange(word=reprise /reprise/, score=None, ranges=((224, 239),)),                                                                                                                         
 WordScoreRange(word=de /de/, score=None, ranges=((241, 244),)),            
 WordScoreRange(word=la /la/, score=None, ranges=((246, 248),)),              
 WordScoreRange(word=coupe /coupe/, score=None, ranges=((250, 258),)),
 WordScoreRange(word=du /du/, score=None, ranges=((259, 262),)),            
 WordScoreRange(word=monde /monde/, score=None, ranges=((264, 272),)),
 WordScoreRange(word=de //, score=None, ranges=((273, 275),)),
 WordScoreRange(word=ski /ski/, score=None, ranges=((278, 289),)),
 WordScoreRange(word=les /les/, score=None, ranges=((295, 299),)),
 WordScoreRange(word=images /images/, score=None, ranges=((303, 318),)),
 WordScoreRange(word=de /de/, score=None, ranges=((321, 323),)),
 WordScoreRange(word=glisse /glisse/, score=None, ranges=((327, 362),)),
 WordScoreRange(word=retrouvent /retrouvent/, score=None, ranges=((375, 394),)),
 WordScoreRange(word=une /une/, score=None, ranges=((398, 403),)),
 WordScoreRange(word=place /place/, score=None, ranges=((409, 424),)),
 WordScoreRange(word=de /de/, score=None, ranges=((430, 432),)),
 WordScoreRange(word=choix /choix/, score=None, ranges=((437, 448),)),
 WordScoreRange(word=dans /dans/, score=None, ranges=((450, 456),)),
 WordScoreRange(word=les /les/, score=None, ranges=((458, 462),)),
 WordScoreRange(word=pages /pages/, score=None, ranges=((465, 479),)),
 WordScoreRange(word=sportives /sportives/, score=None, ranges=((487, 507),)),
 WordScoreRange(word=de /de/, score=None, ranges=((509, 511),)),
 WordScoreRange(word=vos /vos/, score=None, ranges=((513, 519),)),
 WordScoreRange(word=journaux /journaux/, score=None, ranges=((521, 533),)),
 WordScoreRange(word=télévisés /télévisés/, score=None, ranges=((535, 562),)),
 WordScoreRange(word=deux /deux/, score=None, ranges=((568, 576),)),
 WordScoreRange(word=épreuves /épreuves/, score=None, ranges=((581, 618),)),
 WordScoreRange(word=aujourd'hui /aujourd'hui/, score=None, ranges=((629, 654),)),
 WordScoreRange(word=on /onoro/, score=None, ranges=((662, 680),)),
 WordScoreRange(word=a /am/, score=None, ranges=((685, 699),)),
 WordScoreRange(word=santa /santa/, score=None, ranges=((703, 726),)),
 WordScoreRange(word=caterina /qaterina/, score=None, ranges=((729, 761),)),
 WordScoreRange(word=premier /prmier/, score=None, ranges=((768, 783),)),
 WordScoreRange(word=salon /salom/, score=None, ranges=((785, 803),)),
 WordScoreRange(word=géant /géant/, score=None, ranges=((808, 820),)),
 WordScoreRange(word=de /de/, score=None, ranges=((822, 824),)),
 WordScoreRange(word=la /l/, score=None, ranges=((826, 829),)),
 WordScoreRange(word=coupe /coupe/, score=None, ranges=((833, 842),)),
 WordScoreRange(word=du /d/, score=None, ranges=((843, 846),)),
 WordScoreRange(word=monde /mond/, score=None, ranges=((850, 858),))]
```

We can see that before this PR, there are 17 words where timesteps are too early (about one letter shift, it is visible at the end but not at the begining of words because I have trimed spaces).
After this PR, the fit is almost prefect. For some reason, there are still 3 remaining errors, all in the 4 last words.

reuben commented:
@godefv thanks for the PR! I'll review it today.

@utunga you might be interested in trying these changes out and seeing how they affect your timings.

utunga commented:
Thanks @reuben I'm very interested... thanks

reuben commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -96,24 +96,52 @@ DecoderState::next(const double *probs,
+        if (prefix->score == -NUM_FLT_INF) {
+          continue;
+        }
+		if (!prefix->is_empty() && prefix->timesteps.empty()) {
```
Here and throughout: please use spaces and fix the indentation to match the rest of the file.

godefv replied:
Oh, I forgot to check that, I will fix it.

reuben commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -96,24 +96,52 @@ DecoderState::next(const double *probs,
         if (full_beam && log_prob_c + prefix->score < min_cutoff) {
           break;
         }
+        if (prefix->score == -NUM_FLT_INF) {
+          continue;
+        }
```
This seems unrelated to the rest of the PR.

godefv replied:
I had crashes without it in my early tests, I will try it again.

reuben commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -96,24 +96,52 @@ DecoderState::next(const double *probs,
+          continue;
+        }
+		if (!prefix->is_empty() && prefix->timesteps.empty()) {
+		  // This should never happen. But we report it if it does.
+		  std::cerr<<"error: non-empty prefix has empty timestep sequence"<<std::endl;
+		  continue;
```
If this should never happen we should just assert it instead of printing and continuing. That way we find out about problems early.

godefv replied:
ok, sure

reuben commented on native_client/ctcdecode/path_trie.cpp:
```diff
@@ -102,20 +99,18 @@ PathTrie* PathTrie::get_path_trie(unsigned int new_char, unsigned int new_timest
-    parent->get_path_vec(output, timesteps);
+std::vector<unsigned int> PathTrie::get_path_vec() {
+  if (parent == nullptr) {
+	return std::vector<unsigned int>{};
   }
+  std::vector<unsigned int> output_tokens=parent->get_path_vec();
   if (character != ROOT_) {
-    output.push_back(character);
-    timesteps.push_back(timestep);
+  	output_tokens.push_back(character);
   }
+  return output_tokens;
 }
```
This change is causing a new vector to be constructed at each step in the recursion chain just to hold a single character. Taking the output vector as a parameter makes callers a bit more awkward but avoids this entirely, so please keep it like that.

godefv replied:
Copy elision should cause only one vector to be used. Copy elision is even guaranteed by the C++ standard since C++20.
Using return values to return values is more natural.
However, both way would work, so if you still prefer the old way, I will change it.

lissyx replied:
I have not had a careful review, but please keep in mind some of that might be compiled as C++11 or C++14

godefv replied:
After checking, I see that it is actually guaranteed since C++17, but not with named objects. 
I don't expect this change to make any significant difference even with C++11, and I still believe it is better to return by value.
However, it is completely unrelated to this PR, and you both seem to prefer the way it was done before, so I have reverted this unnecessary change.

reuben commented on native_client/ctcdecode/path_trie.h:
```diff
@@ -65,7 +63,10 @@ class PathTrie {
   unsigned int character;
-  unsigned int timestep;
+  std::vector<unsigned int> timesteps;
+  // `timesteps_cur` is a temporary storage for each decoding step. 
+  // At the end of a decoding step, it is moved to `timesteps`.
+  std::vector<unsigned int> timesteps_cur; 
```
Hm, this feels like it's easy to make mistakes without realizing in the code that handles the PathTrie, but I don't have concrete suggestions for now. Gonna think about this while I do more tests.

godefv replied:
It is really the same idea as `log_prob_nb_cur` and `log_prob_b_cur`.
The reason I have not named `timesteps` `timesteps_prev` is that it is used outside the decoding loop directly, like `score` on the probabilities side.

reuben reviewed (commented):
Thanks, your idea makes sense and I think it moves us closer to the direction of having more accurate timings. I need to do a more careful analysis of the main loop changes and do some testing around performance and memory usage, so it'll take me some more time. In the mean time here's some preliminary comments.

godefv commented on native_client/ctcdecode/path_trie.cpp:
```diff
@@ -102,20 +99,16 @@ PathTrie* PathTrie::get_path_trie(unsigned int new_char, unsigned int new_timest
+    parent->get_path_vec(output);
     output.push_back(character);
-    timesteps.push_back(timestep);
   }
```
Here, I realise that I have not completely reverted to the way it was.
It seems to me that it does exactly the same thing with less code, though.
Tell me if you want me to put it back exactly as it was.

reuben commented:
I've benchmarked this PR with the 0.8.2 checkpoint and the Librispeech test clean dataset, by doing test epochs. Unfortunately, there's a significant performance regression, the epoch time went from 240s average to 360s with this PR applied. I think the idea is sound and does get us better timings, but the implementation needs to be different to avoid trashing memory with copying the entire timesteps vector for all prefixes.

godefv commented:
Thanks for the feedback. 
There were indeed too many copies of timesteps vectors, so I have improved on that point in my last commit.
But I have only tested single inference. Could you run your performance test again with these last changes ?

reuben commented:
> Could you run your performance test again with these last changes ?

It speeds it up a bit, to 324s per run.

reuben commented:
I suspect in order to minimize the regression, the timesteps sequence itself needs to be a tree structure as well. Even only copying the sequences once per time step that's still a lot of redundant data copying over the whole decoding process. If the timesteps sequence is an independent tree with PathTrie pointing to its nodes, then it can be walked up to obtain the final sequence at decoding end, and there's no heavy allocation and copying of timesteps vectors during decoding.

godefv commented:
I agree...
I had thought about making a second tree as well, but I was hoping that the simpler change I have made would be enough.

godefv commented:
I have implemented the timestep tree. There are still two potential issues :
1. There should be less copies of timesteps but the number of dynamic allocation will not be lower, because I have replaced each vector copy by a dynamic allocation of a new timestep tree node. This can be solved by using an object pool. 
2. Obsolete data in the timestep tree is not removed, which might cause too much memory to be used. To solve this, I could remove the `children` field and use a vector of timestep nodes at current timestep instead (there is no forward traversal of the timestep tree). Then, `shared_ptr`s would destroy unreferenced data when `remove()` is called on the `PathTrie` nodes.

lissyx commented:
> I have implemented the timestep tree. There are still two potential issues :
> 
>     1. There should be less copies of timesteps but the number of dynamic allocation will not be lower, because I have replaced each vector copy by a dynamic allocation of a new timestep tree node. This can be solved by using an object pool.
> 
>     2. Obsolete data in the timestep tree is not removed, which might cause too much memory to be used. To solve this, I could remove the `children` field and use a vector of timestep nodes at current timestep instead (there is no forward traversal of the timestep tree). Then, `shared_ptr`s would destroy unreferenced data when `remove()` is called on the `PathTrie` nodes.

Have you been able to run figures to compare the impact before / after ? That would really help us!

lissyx commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -36,6 +36,8 @@ DecoderState::init(const Alphabet& alphabet,
   prefix_root_.reset(root);
   prefixes_.push_back(root);
+  
+  timestep_tree_root_=std::make_shared<TimestepTreeNode>(nullptr, 0);
```
nit: spaces

lissyx commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -96,7 +98,7 @@ DecoderState::next(const double *probs,
           break;
         }
-        assert(prefix->is_empty() || !prefix->timesteps.empty());
+        assert(prefix->is_empty() || prefix->timesteps!=nullptr);
```
nit: spaces

lissyx commented on native_client/ctcdecode/path_trie.cpp:
```diff
@@ -172,8 +172,16 @@ void PathTrie::iterate_to_vec(std::vector<PathTrie*>& output) {
+      }
+      if (timesteps == nullptr){
+          timesteps = add_child(previous_timesteps, new_timestep);
+      }
```
nit: spaces everywhere

godefv commented:
On my samples, which are inferences on one file at a time, the current code is roughly 10% faster than before the timestep tree. But I have some overhead around so it might be better on your test set.

I have done some research on shared_ptr and object pools yesterday, and I need some feedback before going further. I will fix spaces when we agree on the implementation. 

I have not tried the two other improvements that I have suggested. Because there is no standard object pool, and last time I checked, open source implementations have issues. An object pool providing `std::shared_ptr` is even more problematic, because `std::shared_ptr` dynamically allocates a control block. Allocating the object in an object pool and leaving the control block dynamically allocated does not reduce the number of dynamic allocations compared to using `std::make_shared`. Unfortunately, the standard library does not expose the type of the control block, so that we cannot create an object pool of "control block+object".

So, having both the object pool and the shared_ptr requires to reimplement a shared_ptr and an object pool, which represents some work. 

I would like to know, with the current code, on your test set, if the speed should be improved more and if the memory usage is acceptable.
If only the speed should be improved, I can use `unique_ptr` and an object pool that I have or one of your choice.
If only the memory should be improved, I can fix it quite easily the way I said.
If both should be improved, do you agree that I need a custom shared_ptr and a custom object pool ?

reuben commented:
This version is the fastest one yet, averaging a LibriSpeech-test-clean test epoch in 272s, which is a 13% slowdown from current master.

reuben commented:
I'm doing some memory profiling but it's going very slowly, will report results once I have them.

godefv commented:
Ok, good.

I have just tried the object pool I have, and it seems that it speeds up again by about 6% on my samples.

godefv commented:
Just to be clear, my object pool provides unique_ptr that I move to shared_ptr. So, the control block is still allocated dynamically (maybe the standard library implementation pools them internaly ?).

reuben commented:
> maybe the standard library implementation pools them internaly ?

In any case that could vary across standard libraries. Instead of using a shared_ptr can't you just do a full pool deallocation at the end of a stream?

godefv commented:
If the memory usage is acceptable (you will tell me), I will use `unique_ptr` instead `shared_ptr` and deallocate at the end of each stream.

godefv commented:
@reuben Hi, have you got a picture of the memory usage ? I have the impression that it is ok on my samples.

On my side, I have tested the object pool with unique pointers stored in the timestep tree. It is 13% faster on my samples than the previous version in this PR (and normally the same memory usage). So, in terms of speed, I expect that it should be roughly equivalent to the current master.

The memory is recycled by the object pool when the PathTrie is destroyed, so it should not leak.

godefv commented:
Oh, and I have just copy-pasted my object pool from another project. I don't know if you want something special about it. 
You can change namespaces or whatever you don't like.

reuben commented:
Yes, sorry, massif took so long to finish that I forgot to comment here. Memory usage looks OK with ec55597 but there was still a significant runtime regression. Testing the latest version now.

reuben commented:
I'm getting a segfault on the first test batch.

reuben commented:
[Switching to Thread 0x7fffa578f700 (LWP 46388)]
0x00007ffef96ae615 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
(gdb) bt
#0  0x00007ffef96ae615 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#1  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#2  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#3  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#4  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#5  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#6  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#7  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#8  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#9  0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#10 0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#11 0x00007ffef96ae624 in PathTrie::iterate_to_vec(std::vector<PathTrie*, std::allocator<PathTrie*> >&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#12 0x00007ffef96b633c in DecoderState::next(double const*, int, int) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#13 0x00007ffef96b7f19 in ctc_beam_search_decoder(double const*, int, int, Alphabet const&, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long) ()
   from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#14 0x00007ffef96bae46 in std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<std::vector<Output, std::allocator<Output> > >, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<std::_Bind<std::vector<Output, std::allocator<Output> > (*(double const*, int, int, Alphabet, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long))(double const*, int, int, Alphabet const&, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long)>, std::allocator<int>, std::vector<Output, std::allocator<Output> > ()>::_M_run()::{lambda()#1}, std::vector<Output, std::allocator<Output> > > >::_M_invoke(std::_Any_data const&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#15 0x00007ffef96b9229 in std::__future_base::_State_baseV2::_M_do_set(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*) ()
   from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#16 0x00007ffff75d2827 in __pthread_once_slow (once_control=0x4d94938, init_routine=0x7fff6c6f58a0 <__once_proxy>) at pthread_once.c:116
#17 0x00007ffef96b9c5c in std::_Function_handler<void (), std::future<std::result_of<std::vector<Output, std::allocator<Output> > (&(double const*, int const&, int&, Alphabet const&, unsigned long&, double&, unsigned long&, std::shared_ptr<Scorer>&, unsigned long&))(double const*, int, int, Alphabet const&, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long)>::type> ThreadPool::enqueue<std::vector<Output, std::allocator<Output> > (&)(double const*, int, int, Alphabet const&, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long), double const*, int const&, int&, Alphabet const&, unsigned long&, double&, unsigned long&, std::shared_ptr<Scorer>&, unsigned long&>(std::vector<Output, std::allocator<Output> > (&)(double const*, int, int, Alphabet const&, unsigned long, double, unsigned long, std::shared_ptr<Scorer>, unsigned long), double const*&&, int const&, int&, Alphabet const&, unsigned long&, double&, unsigned long&, std::shared_ptr<Scorer>&, unsigned long&)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#18 0x00007ffef96b9900 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<ThreadPool::ThreadPool(unsigned long)::{lambda()#1}> > >::_M_run() ()
   from /tmp/venv/lib/python3.6/site-packages/ds_ctcdecoder/_swigwrapper.cpython-36m-x86_64-linux-gnu.so
#19 0x00007fff6c6f66df in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#20 0x00007ffff75ca6db in start_thread (arg=0x7fffa578f700) at pthread_create.c:463
#21 0x00007ffff7903a3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

reuben commented:
With a debug build I get tons of assertion failures before the crash:

```
python: ctc_beam_search_decoder.cpp:99: void DecoderState::next(const double*, int, int): Assertion `prefix->is_empty() || prefix->timesteps != nullptr' failed.
```

reuben commented:
Ah, no, just one per thread, of course.

godefv commented:
Unfortunately, it works on my side with my files.
It seems that you have this issue with all your files, whereas I have it with none of mine. So, it may not depend on the the file used.
What compiler do you use ?

reuben commented:
Can you try building the decoder package and running a test set? From what I understand you're testing with the deepspeech binary rather than by doing a test epoch. You should be able to just do this, from the root of the repo:

```
cd native_client/ctcdecode && (make bindings-debug NUM_PROCESSES=12 && pip install -U dist/*.whl); cd ../..
./bin/run-ldc93s1.sh --epochs 1
```

And it should reproduce the failure. By the way, this warning might be relevant:

```
In file included from path_trie.cpp:1:
In file included from ./path_trie.h:12:
third_party/object_pool/object_pool.h:63:28: warning: field 'free_object_slots_end' is uninitialized when used here [-Wuninitialized]
                free_object_slots_begin{ free_object_slots_end }, // At the begining, set the 2 iterators at the same value to simulate a full pool.
                                         ^
./path_trie.h:112:63: note: in instantiation of member function 'godefv::memory::object_pool_t<TreeNode<unsigned int>, allocator, 1024>::object_pool_t' requested here
    static godefv::memory::object_pool_t<TreeNode<NodeDataT>> tree_node_pool;
                                                              ^
path_trie.cpp:183:23: note: in instantiation of function template specialization 'add_child<unsigned int, unsigned int &>' requested here
          timesteps = add_child(previous_timesteps, new_timestep);
                      ^
1 warning generated.
```

godefv commented:
Hum, it says :
```
 ./bin/run-ldc93s1.sh --epochs 1
+ [ ! -f DeepSpeech.py ]
+ [ ! -f data/ldc93s1/ldc93s1.csv ]
+ echo Downloading and preprocessing LDC93S1 example data, saving in ./data/ldc93s1.
Downloading and preprocessing LDC93S1 example data, saving in ./data/ldc93s1.
+ python -u bin/import_ldc93s1.py ./data/ldc93s1
Traceback (most recent call last):
  File "bin/import_ldc93s1.py", line 7, in <module>
    from mozilla_voice_stt_training.util.downloader import maybe_download
ImportError: No module named mozilla_voice_stt_training.util.downloader
```

godefv commented:
Are you using python 2 or python 3 ?

godefv commented:
It seems that the DeepSpeech code is using `pip` and `python` to run scripts everywhere, but it actually requires python3 (I got "deepspeech-training requires Python '>=3.5, <4' but the running Python is 2.7.17"). So, I solved the above error by replacing `python` by `python3`

godefv commented:
Fixed ! 

After all, unrelated code was not so unrelated, and it indeed prevents crashes.

The reason is that the algorithm from master does not prune paths with zero probability. When such a path is being extended with a character which also has a probability of zero, then the following code is skipped and `timesteps` remains `nullptr`:
```
          if (prefix_new->log_prob_nb_cur < log_p) {
            // record data needed to update timesteps
            // the actual update will be done if nothing better is found
            prefix_new->previous_timesteps = prefix->timesteps;
            prefix_new->new_timestep = abs_time_step_;
          }
```
I could hve fixed it by adding a condition like ` || (refix_new->log_prob_nb_cur == -NUM_FLT_INF && prefix_new->timesteps == nullptr` but it seems quite artificial to support all those paths with zero probability.
Moreover, the check I have put back have prevented crashes twice already, and it looks like a very reasonable check.

reuben commented:
Still crashing on this LibriSpeech file with the 0.8.2 checkpoint:
[8455-210777-0000.wav.zip](https://github.com/mozilla/DeepSpeech/files/5224087/8455-210777-0000.wav.zip)

reuben commented:
That was just the first file in the CSV, I've tested a few others and most of them crashed :(

godefv commented:
Hum, ok, I will look at it.

godefv commented:
Could you give me the command you have used ?
If I use `./deepspeech --model deepspeech-0.8.1-models.pbmm --scorer deepspeech-0.8.1-models.scorer --audio /tmp/8455-210777-0000.wav`, it does not use the decoder I have compiled and installed.
If I use `transcribe.py`, I don't know how to specify what model to use.

reuben commented:
`python DeepSpeech.py --checkpoint_dir deepspeech-0.8.2-checkpoint --n_hidden 2048 --one_shot_infer path/to/audio.wav`

godefv commented:
Ok, thanks, `deepspeech-0.8.2-checkpoint.tar.gz` is indeed available in addition to `deepspeech-0.8.2-models.pbmm`

godefv commented:
Well, sorry about those bugs. I hope it is good now.
The root of the timestep tree was not handled properly and it also forced the first timestep of each path to be 0 (which was wrong).

On my samples, The first timestep is always zero because my CTC loss converged that way. So, I didn't see any issue.

reuben commented:
Still crashing. I'll zip the files so you can test it.

reuben commented:
[bad_batch.zip](https://github.com/mozilla/DeepSpeech/files/5230631/bad_batch.zip)
```python DeepSpeech.py --checkpoint_dir deepspeech-0.8.2-checkpoint --n_hidden 2048 --test_files bad_batch/bad_batch.csv --test_batch_size 20```

godefv commented:
One file at a time, it works.
Of course, my object pool is not thread safe, so...
thread_local object pool solved this !

reuben commented:
It works! I've opened a PR to run all tests here: https://github.com/mozilla/DeepSpeech/pull/3325

reuben commented on native_client/ctcdecode/path_trie.h:
```diff
@@ -9,6 +9,28 @@
+TreeNode<NodeDataT>* add_child(TreeNode<NodeDataT>* node, ChildDataT&& data_);
+
+template<class DataT>
+std::vector<DataT> get_history(TreeNode<DataT> const*, TreeNode<DataT> const* = nullptr);
```
Please name the parameters and briefly document them, just a couple lines is OK.

reuben reviewed (commented):
A few nits! Overall looks great.

godefv commented on native_client/ctcdecode/ctc_beam_search_decoder.cpp:
```diff
@@ -205,11 +239,13 @@ DecoderState::decode(size_t num_results) const
+    output.timesteps  = get_history(prefix->timesteps);
+    output.confidence = scores[prefix];
     outputs.push_back(output);
+    if (outputs.size() >= num_returned) break;
```
I believe the opposite is true. 

The range-based for loop has been introduced to remove code like `prefixes_copy[i]`. Such code is a major cause of crashes, so every time I read this I have to wonder "Can this code crash ?". This makes it less readable. 
Then, I go back to the loop statement and read that `i` can go up to `num_returned-1`. I don't know, just by reading this line, whether this value is less than `prefixes_copy.size()`. So, I have to go go back again to check that. Again less readable.
In addition, it is possible that someone changes the definition of `num_returned` and it may cause a crash later in the loop.

In contrast, range-based for loops almost never crash. People used to complain about C/C++ segfaulting too often. Modern C++ has removed most causes of segfaults, and that's why ranged based for loops are almost always preferred.

Regarding the condition + break, it makes it clear that `outputs` is limited to a maximum size of `num_returned`, which is what the name "num_returned" means. With previous code, it was slightly less obvious.

reuben replied:
In C++ land you should be thinking "can this code crash?" all the time, regardless of using array indices or not. As evidenced by the last few commits in this PR. Modern C++ is a thin veil over the same crashy stuff. If someone changes the `num_returned` logic in your version, the code won't crash, but it'll silently do the wrong thing, so you need to be equally vigilant. In any case, I don't care enough about this case to force it, but watch out for buying more modernness than "Modern C++" can actually offer.

godefv replied:
We had quite a few crashes indeed ! 
It was due to null pointers, and thread safety, which are not covered well by "Modern C++" (smart pointers and thread_local help, but they don't prevent accessing bad memory).

Anyway, you have confirmed that you prefer like it was before, and this change is unrelated to this PR, so I should revert it.

reuben commented:
> It works! I've opened a PR to run all tests here: #3325

All green!

godefv commented:
Great, I will look after spaces and documentation.

godefv commented:
Regarding the object pool, I have published it there https://github.com/godefv/memory.
For now, you have your own copy and can do whatever you like with it. 

In my version, I have removed the namespace `memory::`, which is not very useful.

reuben commented:
Oh, that's a good point I forgot to mention. Could you add a `third_party/object_pool/README.mozilla` text file saying basically "This code was imported from this repository on this date, commit SHA1. It's licensed under X license." (And you should probably add licensing details to your repo too!).

godefv commented:
Ok. I have wondered a few times what licence I would choose. 
I think I like Creative Commons licenses, and this one in particular CC BY-SA.

From their website, it says :
> This license lets others remix, adapt, and build upon your work even for commercial purposes, as long as they credit you and license their new creations under the identical terms. This license is often compared to “copyleft” free and open source software licenses. All new works based on yours will carry the same license, so any derivatives will also allow commercial use. This is the license used by Wikipedia, and is recommended for materials that would benefit from incorporating content from Wikipedia and similarly licensed projects. 

I am open to choose something else if needed. Do you have any advice on this point ?

godefv commented:
I see that DeepSpeech has "Mozilla Public License 2.0". I don't know what license is compatible with what...
Obviously, I need to choose something compatible with this one.

lissyx commented:
> Ok. I have wondered a few times what licence I would choose.
> I think I like Creative Commons licenses, and this one in particular CC BY-SA.
> 
> From their website, it says :
> 
> > This license lets others remix, adapt, and build upon your work even for commercial purposes, as long as they credit you and license their new creations under the identical terms. This license is often compared to “copyleft” free and open source software licenses. All new works based on yours will carry the same license, so any derivatives will also allow commercial use. This is the license used by Wikipedia, and is recommended for materials that would benefit from incorporating content from Wikipedia and similarly licensed projects.
> 
> I am open to choose something else if needed. Do you have any advice on this point ?

I'm unsure this is going to mix well with the rest of the codebase, though, but I'm no license expect.

godefv commented:
I am interested in finding a good license for my projects in general, so I am looking into it.
If you don't know in advance what is compatible, don't waste your time, I will find it.

godefv commented:
Well, for this one, I guess the CC0 (public domain) will be good.

godefv commented:
Licensing is done. I have assumed that the CC0 license allows you to do more or less anything without worrying if it is allowed or not.

reuben reviewed (approved):
LGTM. @lissyx do you want to take another look or good to merge?

lissyx commented:
Honestly, you seems to have nailed it pretty well, let's not hold it too long. Maybe good to do a new alpha release with that ?

godefv commented:
Cool ! 
This was quick, with useful feedback (testing) and constructive comments, leading to several improvements compared to the beginning. I am quite happy with that. Thanks !

reuben commented:
Thanks for the excellent contribution and for keeping up with the review process, I know it can be tiring sometimes!

reuben commented:
Should be published as 0.9.0-alpha.9
